recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb

```diff
@@ -0,0 +1,505 @@
+        "def formatting_prompts_func(example):\n",
+        "    output_texts = []\n",
+        "    for i in range(len(example['question'])):\n",
+        "        text = f\"<|system|>\\nYou are a helpful assistant\\n<|user|>\\n{example['question'][i]}\\n<|assistant|>\\n{example['answer'][i]}<|endoftext|>\"\n",
+        "        output_texts.append(text)\n",
+        "    return output_texts\n",
+        "\n",
+        "response_template = \"\\n<|assistant|>\\n\"\n",
+        "\n",
+        "from trl import DataCollatorForCompletionOnlyLM\n",
+        "\n",
+        "response_template_ids = tokenizer.encode(response_template, add_special_tokens=False)[2:]\n",
+        "collator = DataCollatorForCompletionOnlyLM(response_template_ids, tokenizer=tokenizer)\n",
+        "\n",
+        "\n",
+        "# Apply qLoRA\n",
+        "qlora_config = LoraConfig(\n",
+        "    r=16,  # The rank of the Low-Rank Adaptation\n",
+        "    lora_alpha=32,  # Scaling factor for the adapted layers\n",
+        "    target_modules=[\"q_proj\", \"v_proj\"],  # Layer names to apply LoRA to\n",
+        "    lora_dropout=0.1,\n",
+        "    bias=\"none\"\n",
+        ")\n",
+        "\n",
+        "# Initialize the SFTTrainer\n",
+        "training_args = TrainingArguments(\n",
+        "    output_dir=\"./results\",\n",
+        "    evaluation_strategy=\"epoch\",\n",
+        "    learning_rate=2e-4,\n",
+        "    per_device_train_batch_size=6,\n",
+        "    per_device_eval_batch_size=6,\n",
+        "    num_train_epochs=3,\n",
+        "    logging_dir='./logs',\n",
+        "    logging_steps=10,\n",
+        "    fp16=True\n",
+        ")\n",
+        "\n",
+        "max_seq_length = 250\n",
+        "\n",
+        "trainer = SFTTrainer(\n",
+        "    model=model,\n",
+        "    args=training_args,\n",
+        "    train_dataset=split_dataset['train'],\n",
+        "    eval_dataset=split_dataset['test'],\n",
+        "    tokenizer=tokenizer,\n",
+        "    peft_config = qlora_config,\n",
+        "    formatting_func=formatting_prompts_func,\n",
+        "    data_collator=collator,\n",
+        "    max_seq_length=max_seq_length,\n",
+        ")\n",
+        "\n",
+        "training_setup_loadtime = timeit.default_timer() - start_time\n"
+      ],
+      "metadata": {
+        "id": "Ng0W0ULxsE_i"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "## Training Process  \n",
+        "  \n",
+        "With all the preparations complete, we now start the training process. The model will be exposed to numerous examples from our legal dataset, gradually adjusting its understanding of legal concepts.  \n",
+        "  \n",
+        "We'll monitor the training loss over time, which should decrease as the model improves its performance on the task. After training, we'll save the fine-tuned model for future use.  "
```

Comment:
Which numbers show the training loss?

Reply:
I see them now; they appear after the first epoch. Maybe mention that here.

Reply:
It's a fair call out- like i said on the call, I'm trying to avoid going down rabbit holes- if you do you end up with a monstrous notebook, that still isn't high enough quality to use for real fine-tuning, but is very intimidating. I think I'll turn up how often it reports the loss in the logger and add one text block about over fitting and early stopping and then link out to more resources.

Reply:
Yeah, I get that you don't want to explain the loss function here. I was really just asking for clarity on where I can see the actual output of the loss numbers (since the recipe says we will "monitor them over time"), b/c they didn't show up right away. Maybe just mention that the first loss numbers will show up after the first epoch is trained.

Reply:
That's a good note. Ill update it to spit out more training info- I also see how I misread your initial comment.